### PR TITLE
Expand literate Haskell semantics

### PR DESCRIPTION
**Pull request by Gabriel439**
… to include:

* Binary encoding
* α-normalization
* β-normalization
* Equivalence checking

Related to: https://github.com/dhall-lang/dhall-lang/issues/959

Note that this does not include binary decoding.  I was mainly
focused on completing all the dependencies for equivalence checking,
and binary decoding was not necessary for that, yet.

### CODE REVIEW

File: standard/Equivalence.lhs-boot [hunk 1]
```diff
@@ -0,0 +1,7 @@
+```haskell
+module Equivalence where
+
+import Syntax (Expression)
+
+equivalent :: Expression -> Expression -> Bool
+```
```
Comment (Nadrieril) [inline standard/Equivalence.lhs-boot:7]: What's this file for?

Comment (Gabriel439) [reply standard/Equivalence.lhs-boot:7]: Haskell requires these types of file to break import cycles.  See this for more details:

* [GHC manual - How to compile mutually recursive modules](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/separate_compilation.html#how-to-compile-mutually-recursive-modules

**Comment by Nadrieril** [comment]
I didn't read through all of it but it looks very clean. Were you able to read this at all? It seems we don't have a parser yet.

**Comment by Gabriel439** [comment]
@Nadrieril: I haven't actually tested that the code works, yet.  I can begin work on the parser for the next step.

The main reason I focused on this part first is that I wanted to flush out some non-trivial code (e.g. the β-normalization logic) to make sure that people are okay with the coding style before I proceed further.

File: standard/beta-normalization.md [hunk 2]
```diff
@@ -323,6 +458,39 @@ numbers:
...
+
+        let t₁  | m == 0 =
+                    betaNormalize b
+                | otherwise =
+                    betaNormalize
+                        (Application g
+                            (Application
+                                (Application
+                                    (Application
+                                        (Application (Builtin NaturalFold)
```
Comment (Nadrieril) [inline standard/beta-normalization.md:481]: This pile of `Application`s is a bit ugly. Could we do
```haskell
let (@) = Application
in  betaNormalize (g @ (NaturalFold @ (NaturalLiteral (m - 1)) @ _B @ g @ b))
```
or somethinglike that?

**Review by Nadrieril — APPROVED** [review]
Thanks for the impressive work! Looks good to me. Normalization of builtins can be hard to follow, but I'm not sure we can do much better anyways

**Comment by Gabriel439** [comment]
@Nadrieril: I took a stab at your suggestion by using GHC's support for pattern synonyms, like this:

```haskell
{-| Convenient synonym for `Application` so that β-normalization code is easier
    to read
-}
pattern (:@) :: Expression -> Expression -> Expression
pattern f :@ x = Application f x

infixl :@
```

… which does make the code much clearer, but as I was applying the change the type-checking began to slow down significantly until I hit this warning:

```
BetaNormalization.lhs:(103,1)-(2577,68): error: [-Werror]
    Pattern match checker exceeded (2000000) iterations in
    an equation for ‘betaNormalize’. (Use -fmax-pmcheck-iterations=n
    to set the maximum number of iterations to n)
    |
103 | betaNormalize (Constant c) = Constant c
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^...
```

… so I think I'll avoid using the pattern synonym for now until we begin using GHC 8.10 to build the literate Haskell semantics, which appears to have fixed the performance issue (See: https://gitlab.haskell.org/ghc/ghc/-/issues/11822)

File: standard/Equivalence.lhs-boot (hunk 1 above)
Comment (sjakobi) [reply standard/Equivalence.lhs-boot:7]: How about adding a comment that preserves the answer to this question?
